gt/main.c

```diff
@@ -227,10 +227,194 @@ fill_eth_hdr(struct rte_mbuf *m,
 	return 0;
 }
 
+static void
+fill_eth_hdr_reverse(struct ether_hdr *eth_hdr,
+	struct gt_packet_headers *pkt_info)
+{
+	struct ether_hdr *raw_eth = (struct ether_hdr *)pkt_info->l2_hdr;
+	ether_addr_copy(&raw_eth->s_addr, &eth_hdr->d_addr);
+	ether_addr_copy(&raw_eth->d_addr, &eth_hdr->s_addr);
+	eth_hdr->ether_type = rte_cpu_to_be_16(pkt_info->outer_ip_ver);
+}
+
+static struct rte_mbuf *
+alloc_and_fill_notify_pkt(unsigned int socket, struct ggu_policy *policy,
+	struct gt_packet_headers *pkt_info, struct gt_config *gt_conf)
+{
+	uint8_t *data;
+	uint16_t ethertype = pkt_info->outer_ip_ver;
+	struct ether_hdr *notify_eth;
+	struct ipv4_hdr *notify_ipv4;
+	struct ipv6_hdr *notify_ipv6;
+	struct udp_hdr *notify_udp;
+	struct ggu_common_hdr *notify_ggu;
+
+	struct rte_mbuf *notify_pkt = rte_pktmbuf_alloc(
+		gt_conf->net->gatekeeper_pktmbuf_pool[socket]);
+	if (notify_pkt == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gt: failed to allcate notification packet!");
+		return NULL;
+	}
+
+	if (ethertype == ETHER_TYPE_IPv4) {
+		notify_eth = (struct ether_hdr *)rte_pktmbuf_append(notify_pkt,
+			sizeof(struct ether_hdr) + sizeof(struct ipv4_hdr) +
+			sizeof(struct udp_hdr) + sizeof(struct ggu_common_hdr));
+		notify_ipv4 = (struct ipv4_hdr *)&notify_eth[1];
+		notify_udp = (struct udp_hdr *)&notify_ipv4[1];
+		notify_ggu = (struct ggu_common_hdr *)&notify_udp[1];
+	} else if (ethertype == ETHER_TYPE_IPv6) {
+		notify_eth = (struct ether_hdr *)rte_pktmbuf_append(notify_pkt,
+			sizeof(struct ether_hdr) + sizeof(struct ipv6_hdr) +
+			sizeof(struct udp_hdr) + sizeof(struct ggu_common_hdr));
+		notify_ipv6 = (struct ipv6_hdr *)&notify_eth[1];
+		notify_udp = (struct udp_hdr *)&notify_ipv6[1];
+		notify_ggu = (struct ggu_common_hdr *)&notify_udp[1];
+	} else
+		rte_panic("Unexpected condition: gt fills up a notify packet with unknown ethernet type %hu\n",
+			ethertype);
+
+	/* Fill up the policy decision. */
+	memset(notify_ggu, 0, sizeof(*notify_ggu));
+	notify_ggu->v1 = GGU_PD_VER1;
+	if (pkt_info->inner_ip_ver == ETHER_TYPE_IPv4
```

Comment:
While it's correct to use `pkt_info->inner_ip_ver` here, you really want to use `policy->flow.proto` here since it makes no assumption on the policy and the packet. Please adjust the other occurrences accordingly.